src/irmin-pack/inode.ml

```diff
@@ -402,60 +402,63 @@ struct
 
     let stable t = t.stable
 
-    type acc = {
-      cursor : int;
-      values : (step * value) list list;
-      remaining : int;
-    }
-
-    let empty_acc n = { cursor = 0; values = []; remaining = n }
-
-    let rec list_entry layout ~offset ~length acc = function
-      | None -> acc
-      | Some i -> list_values layout ~offset ~length acc (Ptr.target layout i).v
-
-    and list_tree layout ~offset ~length acc t =
-      if acc.remaining <= 0 || offset + length <= acc.cursor then acc
-      else if acc.cursor + t.length < offset then
-        { acc with cursor = t.length + acc.cursor }
-      else Array.fold_left (list_entry layout ~offset ~length) acc t.entries
-
-    and list_values layout ~offset ~length acc v =
-      if acc.remaining <= 0 || offset + length <= acc.cursor then acc
-      else
-        match v with
-        | Values vs ->
-            let len = StepMap.cardinal vs in
-            if acc.cursor + len < offset then
-              { acc with cursor = len + acc.cursor }
-            else
-              let to_drop =
-                if acc.cursor > offset then 0 else offset - acc.cursor
-              in
-              let vs =
-                StepMap.to_seq vs |> Seq.drop to_drop |> Seq.take acc.remaining
-              in
-              let n = List.length vs in
-              {
-                values = vs :: acc.values;
-                cursor = acc.cursor + len;
-                remaining = acc.remaining - n;
-              }
-        | Tree t -> list_tree layout ~offset ~length acc t
-
-    let list_v layout ?(offset = 0) ?length v =
-      let length =
+    type acc = { mutable acc_offset : int; mutable acc_length : int }
+
+    let rec seq_tree layout acc bucket_seq k () =
+      match bucket_seq () with
+      | Seq.Nil -> k ()
+      | Seq.Cons (None, rest) -> seq_tree layout acc rest k ()
+      | Seq.Cons (Some i, rest) ->
+          let trg = Ptr.target layout i in
+          let length = length trg in
+          if acc.acc_offset - length >= 0 then (
+            (* This branch is optional, it just acts as a shortcut.
+               [seq_values] could handle the pagination by itself. *)
+            acc.acc_offset <- acc.acc_offset - length;
+            seq_tree layout acc rest k ())
+          else seq_v layout acc trg.v (seq_tree layout acc rest k) ()
+
+    and seq_values layout acc value_seq k () =
+      match value_seq () with
+      | Seq.Nil -> k ()
+      | Cons (x, rest) ->
+          if acc.acc_offset = 0 then (
+            acc.acc_length <- acc.acc_length - 1;
+            if acc.acc_length = 0 then Seq.Cons (x, Seq.empty)
```

Comment:
any reason to not call `k` here?

Reply:
Yes, this case corresponds to the end of the requested pagination

Reply:
It would be better to write it in CPS style ; something interesting we could do then is to decide what type should these function return (a seq or a list) without too much duplication

Reply:
It would be better to write it in CPS style ; something interesting we could do then is to decide what type should these function return (a seq or a list) without too much duplication

Reply:
Because of the pagination it is not possible to write true CPS. The continuation has to be dropped here.